### PR TITLE
Hdf5 fill value

### PR DESCRIPTION
**Pull request by jgallagher59701**
This is an early version of the Fill Value fix; I want to see if it passes tests on Linux.

### CODE REVIEW

**Comment by jgallagher59701** [comment]
NB: I don't think we should merge this, but I thought it could do with a review. The code as it is now can lead into the next phase - actually using the fill value info. Note also that support for COMPACT will come in this second phase.

File: modules/dmrpp_module/build_dmrpp_util.cc
Comment (kyang2014) [inline modules/dmrpp_module/build_dmrpp_util.cc:86]: If you use this code, I think you should add a comment that they are adapted from the HDF5 handler . The same for the read_vlen_string().

File: modules/dmrpp_module/build_dmrpp_util.cc
Comment (kyang2014) [inline modules/dmrpp_module/build_dmrpp_util.cc:308]: If this file is commented out. Please ignore my review. Otherwise, I don't understand your logic here. I think you copy the code from an example code I gave to you but didn't do in the correct context. If the fillvalue is not defined, it doesn't mean the storage size is 0.  Maybe this is  only for the debugging purpose. If this is  the case, you may mark it differently.

File: modules/dmrpp_module/build_dmrpp_util.cc
Comment (kyang2014) [inline modules/dmrpp_module/build_dmrpp_util.cc:324]: Why not using vector<char> to replace malloc() and free() ?

File: modules/dmrpp_module/build_dmrpp_util.cc
Comment (kyang2014) [inline modules/dmrpp_module/build_dmrpp_util.cc:619]: No need to handle fillvalue for string

File: modules/dmrpp_module/build_dmrpp_util.cc
Comment (kyang2014) [inline modules/dmrpp_module/build_dmrpp_util.cc:621]: This is even remote. Array datatype is seldom used and also not the fillvalue.

File: modules/dmrpp_module/build_dmrpp_util.cc
Comment (kyang2014) [inline modules/dmrpp_module/build_dmrpp_util.cc:624]: fillvalue doesn't make sense for reference.

File: modules/dmrpp_module/build_dmrpp_util.cc
Comment (kyang2014) [inline modules/dmrpp_module/build_dmrpp_util.cc:625]: Compound is something that may need to be considered later.

**Review by kyang2014 — CHANGES_REQUESTED** [review]
I don't see you handle the empty datasets or empty chunks that are are filled with fillvalue yet.

**Review by kyang2014 — DISMISSED** [review]

File: modules/dmrpp_module/build_dmrpp_util.cc
Comment (jgallagher59701) [reply modules/dmrpp_module/build_dmrpp_util.cc:621]: OK, but we have to handle this because some of our test data does have the fill value attribute set for H5T_ARRAY. It's odd, and I have no idea how to get the value. 

What should I do for this case? clear the fill value so that it does not appear to be set?

File: modules/dmrpp_module/build_dmrpp_util.cc
Comment (jgallagher59701) [reply modules/dmrpp_module/build_dmrpp_util.cc:625]: OK. Maybe I'll make these three cases (COMPOUND, ARRAY and STRING) "unsupported" for now so we have a mnemonic that this needs to be investigated in the future.

File: modules/dmrpp_module/build_dmrpp_util.cc
Comment (kyang2014) [reply modules/dmrpp_module/build_dmrpp_util.cc:621]: Are you sure? Where do you get the data? 

> OK, but we have to handle this because some of our test data does have the fill value attribute set for H5T_ARRAY. It's odd, and I have no idea how to get the value.

Comment (jgallagher59701) [reply modules/dmrpp_module/build_dmrpp_util.cc:621]: You're correct - there are no cases of ARRAY or COMPOUND in the dmrpp builder test data. There are some instances of STRING.

**Comment by jgallagher59701** [comment]
I'm going to leave this PR open and start the next ticket. Thanks for the review work. I think it helped catch some issues before I went charging ahead.

**Review by ndp-opendap — COMMENTED** [review]
Again the `#if 0`'d things make this super hard to understand.

**Review by kyang2014 — COMMENTED** [review]
change to comment since having not reviewed the new code.

**Comment by jgallagher59701** [comment]
Please review. Pay attention to the cc and h files in modules/dmrpp_module. Most of the other parts are baselines and the odd fix I made as problems appeared in the debugger during testing.

File: modules/dmrpp_module/data/fill_value/FValue.h5.dmr
Comment (kyang2014) [inline modules/dmrpp_module/data/fill_value/FValue.h5.dmr:31]: Suggest you add a test to check the expression constraint across filvalue boundary. I think it  should work but a test is better to avoid any future potential issue.

File: modules/dmrpp_module/DMZ.cc
Comment (ndp-opendap) [inline modules/dmrpp_module/DMZ.cc:1188]: // do we need to run this code?
Isn't the test here really for:
            // Are there missing chunks?

File: modules/dmrpp_module/DmrppArray.cc
Comment (ndp-opendap) [inline modules/dmrpp_module/DmrppArray.cc:951]: Is this a style thing you want us all to adopt?

File: modules/dmrpp_module/DmrppArray.cc
Comment (ndp-opendap) [inline modules/dmrpp_module/DmrppArray.cc:956]: ibid

File: modules/dmrpp_module/DmrppCommon.cc
Comment (ndp-opendap) [inline modules/dmrpp_module/DmrppCommon.cc:94]: These calls to ERROR_LOG are crucial to understanding what is happening with this code when it's deployed in the filed, specifically NGAP. I commented them out because things seemed ok, but I think we should leave them in just in case we have issues down the road. They're the right things in the right place and it's not always easy to to sort that out from scratch every time

File: modules/dmrpp_module/DmrppCommon.cc
Comment (ndp-opendap) [inline modules/dmrpp_module/DmrppCommon.cc:98]: ibid

File: modules/dmrpp_module/DmrppCommon.cc
Comment (ndp-opendap) [inline modules/dmrpp_module/DmrppCommon.cc:149]: Why not just do this??  Seems like it should be straight forward...

File: modules/dmrpp_module/DmrppCommon.cc
Comment (ndp-opendap) [inline modules/dmrpp_module/DmrppCommon.cc:297]: This method needs a doxygen (or whatever it is) header.

File: modules/dmrpp_module/DmrppCommon.h
Comment (ndp-opendap) [inline modules/dmrpp_module/DmrppCommon.h:67]: Sweet! It's been a long time since I've seen/used a `union`.

File: modules/dmrpp_module/DmrppCommon.h
Comment (ndp-opendap) [inline modules/dmrpp_module/DmrppCommon.h:235]: Nice catch! That `size_t` business is trouble.

File: modules/dmrpp_module/SuperChunk.cc
Comment (ndp-opendap) [inline modules/dmrpp_module/SuperChunk.cc:79]: I'm don't know what the answer is but it seems like we should resolve this question now or write a ticket to do so.

File: modules/dmrpp_module/SuperChunk.cc
Comment (ndp-opendap) [inline modules/dmrpp_module/SuperChunk.cc:398]: +1

File: modules/dmrpp_module/SuperChunk.h
Comment (ndp-opendap) [inline modules/dmrpp_module/SuperChunk.h:63]: but... What if we end up using a readable name? Will the be useful down the road?

**Review by ndp-opendap — APPROVED** [review]
That's a lot of work @jgallagher59701! Most of it looks great and most of my comments are nitpicky, but I see there are a lot of TODO items, many of which are important to sort out. It seems to me that  we need write follow on tickets for these before we merge this, because once it's back in the main branch it will be harder to remember what needs to be done and more likely that we'll just blow it off to the projects potential detriment.

File: modules/dmrpp_module/DmrppArray.cc
Comment (jgallagher59701) [reply modules/dmrpp_module/DmrppArray.cc:951]: Yes. It's also the default for lots of automated tools.

File: modules/dmrpp_module/DmrppCommon.h
Comment (jgallagher59701) [reply modules/dmrpp_module/DmrppCommon.h:67]: Yeah.

File: modules/dmrpp_module/DmrppCommon.h
Comment (jgallagher59701) [reply modules/dmrpp_module/DmrppCommon.h:235]: Right. site_t is what vector::size() returns, so we need it, but we know our arrays can be bigger. We might switch from writing `unsigned long long` to `uint64_t` now that we have C++11 - it standardizes that while unsigned long long is not _guaranteed_ to be 64-bits.

File: modules/dmrpp_module/SuperChunk.cc
Comment (jgallagher59701) [reply modules/dmrpp_module/SuperChunk.cc:79]: [HYRAX-713](https://bugs.earthdata.nasa.gov/browse/HYRAX-713)

File: modules/dmrpp_module/SuperChunk.h
Comment (jgallagher59701) [reply modules/dmrpp_module/SuperChunk.h:63]: I don't understand - my comment is that we don't need a string here and doing so uses more time than not. An int will be fine and if we need a string lagter on, we do that. 

I'm not changing the code now; this can go in a general TODO ticket and we can revisit the issue.

**Comment by jgallagher59701** [comment]
All of the review comments have been addressed - if this passes the tests now that the #if 0 stuff is gone, I'll merge it and watch as the build progresses. There are a number of new tickets for the various TODO items - mostly in the 'dmr++' sprint (HYRAX-722 to 726).

**Comment by ndp-opendap** [comment]
Thanks James!

**Pull request merged**
